packages/typespec-ts/src/modular/buildOperationUtils.ts

```diff
@@ -0,0 +1,441 @@
+    kind: StructureKind.FunctionOverload,
+    parameters: [
+      {
+        name: "obj",
+        type: typeName
+      }
+    ],
+    docs: functionStatement.docs,
+    returnType: functionStatement.returnType
+  });
+  const implementationParameter = sourceFile
+    .getFunction(functionStatement.name ?? "")
+    ?.getParameters();
+  if (implementationParameter && implementationParameter.length > 0) {
+    const newTypeName =
+      implementationParameter[0]?.getTypeNode()?.getText() + " | " + typeName;
+    implementationParameter[0]?.setType(newTypeName);
+  }
+}
+
+function hasDuplicateFunction(
+  sourceFile: SourceFile,
+  functionStatement: FunctionDeclarationStructure
+) {
+  return sourceFile.getFunctions().some((f) => {
+    const paramTypes = f.getParameters().map((param) => {
+      return param.getName() + param.getType().getText();
+    });
+    const funcParamTypes = functionStatement.parameters?.map((param) => {
+      return param.name + param.type;
+    });
+    return (
+      f.getName() === functionStatement.name &&
+      paramTypes.join().includes(funcParamTypes?.join() ?? "")
+    );
+  });
+}
+
+function deserializeUnionTypesFunction(
+  sourceFile: SourceFile,
+  unionDeserializeTypes: Type[],
+  deserializeFunctionName: string,
+  typeUnionNamesOutput: string | undefined,
+  typeUnionNames: string | undefined
+) {
+  const functionStatement: FunctionDeclarationStructure = {
+    kind: StructureKind.Function,
+    docs: [`deserialize function for ${typeUnionNamesOutput}`],
+    name: deserializeFunctionName,
+    parameters: [{ name: "obj", type: typeUnionNamesOutput }],
+    returnType: typeUnionNames,
+    isExported: true
+  };
+  const statements: string[] = [];
+  for (const type of unionDeserializeTypes) {
+    const functionName = toPascalCase(
+      type.name ??
+        (type.elementType?.name ? type.elementType.name + "Array" : "")
+    );
+    statements.push(
+      `if (is${functionName}(obj)) { return deserialize${functionName}(obj); }`
+    );
+  }
+  statements.push("return obj;");
+  functionStatement.statements = statements.join("\n");
+  if (!hasDuplicateFunction(sourceFile, functionStatement)) {
+    sourceFile.addFunction(functionStatement);
+  }
+}
+
+function getTypePredictFunction(
+  sourceFile: SourceFile,
+  type: Type,
+  typeUnionNames: string | undefined
+): void {
+  if (typeUnionNames === undefined) {
+    return;
+  }
+
+  const statements: string[] = [];
+  if (type.type === "model" && type.name) {
+    const functionStatement: FunctionDeclarationStructure = {
+      kind: StructureKind.Function,
+      docs: [`type predict function fpr ${type.name} from ${typeUnionNames}`],
+      name: `is${toPascalCase(formalizeTypeUnionName(type.name))}`,
+      parameters: [{ name: "obj", type: typeUnionNames }],
+      returnType: `obj is ${type.name}Output`
+    };
+    if (type.properties) {
+      statements.push(
+        `return ${type.properties
+          .filter((p) => !p.optional)
```

Comment:
Does that mean we will pick the first required property as the indicator?

Reply:
No, this is to filter out all the required properties. not sure where you get the first required property ?

Reply:
Hah, got it. I may misunderstand with our generated code because there is only one property involved.

So we will include all required proerties as type indicator, right?

Reply:
we will include all required properties in this type and all its properties of any property which is a model type.  see https://github.com/Azure/autorest.typescript/pull/1962/files#diff-fee0329cc5e0bcbba7157ff4ae985fdbb1f41a2a389e764a084dbdf1487ff82eR533-R543 for details.

Reply:
Could we add a UT that is a special union A | B with their required properties are the same but with different optional parameters?

Reply:
I don't think this is a valid case, if there's an object that is missing the optional properties, we will never know whether it's type A or type B by all means.

Reply:
I may take this case as negative testing, if tsp could define this.